### DISCUSSION TITLE
Debugging Controller Connection Strength

### DISCUSSION THREAD

**Discussion by yo1dog** [darthcloud/BlueRetro#385] [Q&A]
Is there a way to see a controller's connection strength to the ESP32? I was looking through the source and debug flags but didn't see anywhere obvious where the signal strength/RSSI can be obtained.

**Comment by darthcloud** [comment]
You can get the RSSI within the inquiry mode result or by using the HCI Read RSSI command.

See https://www.bluetooth.com/specifications/specs/core-specification-5-3/

page 2140

**Reply by yo1dog** [reply]
Thanks! Here's my quick hack to add polling and printing of RSSI value: https://github.com/yo1dog/BlueRetro/commit/18f64cb583bc6bd52fa8976669c1bccaebf48391

I just setup a task to send the `HCI_Read_RSSI` command every second and print the result. This seems to work. However, I am seeing a range of 0 to -45. When the controller is on the other side of the house, I start seeing packet stalls/loss as expected. However RSSI is being returned as -45 rather than the -90 to -127 range I would expect at that distance.

Docs state:

> For a BR/EDR Controller, the RSSI parameter returns the difference between the measured Received Signal Strength Indication (RSSI) and the limits of the Golden Receive Power Range for a Connection_Handle to another BR/EDR Controller ... any negative value indicates how many dB the RSSI is below the lower limit. ... The lower threshold level corresponds to a received power between -56 dBm and 6 dB above the actual sensitivity of the receiver. ... For an LE transport, the RSSI parameter returns the absolute receiver signal strength value in dBm to ±6 dB accuracy. If the RSSI cannot be read, the RSSI metric shall be set to 127.
> ...
> - RSSI:
>   - LE:
>     - Range: -127 to 20, 127
>     - Units: dBm

Not sure what all that means exactly. But simply adding -56 does not make sense either as then I would read values of -60 when the controller is right next to the ESP32.

Any advice?

**Reply by darthcloud** [reply]
What device are you using? Unless you got a Xbox, Steam controller or a recent mouse/keyboard your device is likely BR/EDR (Classic BT)

BR/EDR
Range: -128 to 127
Units: dB

My understanding is that an ideal signal would be 0dB, <0 wearker, >0 too strong

**Reply by yo1dog** [reply]
Makes sense. This works for me as I only need relative numbers to compare antenna placement anyway.

Any interest in adding this as a debug flag?

**Reply by darthcloud** [reply]
Yes but move it to an existing thread: https://github.com/darthcloud/BlueRetro/blob/10299b69eb992ebd7c3fea69059be2f9df072464/main/bluetooth/host.c#L282

**Reply by yo1dog** [reply]
Will do. Current implementation was a quick hack. I'll look at it more seriously and submit a PR.
